File: benchmarks/tsb/bench_cum_ops.ts
/**
 * Benchmark: cumsum / cumprod / cummax / cummin (Series and DataFrame)
 * Mirrors pandas Series.cumsum(), DataFrame.cumsum(), etc.
 */
import { Series, DataFrame } from "../../src/index.ts";
import {
  cumsum,
  cumprod,
  cummax,
  cummin,
  dataFrameCumsum,
} from "../../src/stats/cum_ops.ts";

const N = 100_000;

// Numeric series for cumsum/cumprod/cummax/cummin
const data = Array.from({ length: N }, (_, i) => (i % 100) + 1);
const series = new Series({ data });

// DataFrame with two columns
const col1 = Array.from({ length: N }, (_, i) => (i % 100) + 1);
const col2 = Array.from({ length: N }, (_, i) => ((i * 3) % 100) + 1);
const df = DataFrame.fromColumns({ a: col1, b: col2 });

const WARMUP = 5;
const ITERS = 20;

// --- warm-up ---
for (let i = 0; i < WARMUP; i++) {
  cumsum(series);
  cummax(series);
  dataFrameCumsum(df);
}

// --- measured: cumsum ---
const t0cs = performance.now();
for (let i = 0; i < ITERS; i++) cumsum(series);
const totalCumsum = performance.now() - t0cs;

// --- measured: cumprod ---
const t0cp = performance.now();
for (let i = 0; i < ITERS; i++) cumprod(series);
const totalCumprod = performance.now() - t0cp;

// --- measured: cummax ---
const t0cx = performance.now();
for (let i = 0; i < ITERS; i++) cummax(series);
const totalCummax = performance.now() - t0cx;

// --- measured: cummin ---
const t0cn = performance.now();
for (let i = 0; i < ITERS; i++) cummin(series);
const totalCummin = performance.now() - t0cn;

// --- measured: dataFrameCumsum ---
const t0df = performance.now();
for (let i = 0; i < ITERS; i++) dataFrameCumsum(df);
const totalDf = performance.now() - t0df;

const total_ms = totalCumsum + totalCumprod + totalCummax + totalCummin + totalDf;
const mean_ms = total_ms / (ITERS * 5);

console.log(
  JSON.stringify({
    function: "cum_ops",
    mean_ms: parseFloat(mean_ms.toFixed(4)),
    iterations: ITERS * 5,
    total_ms: parseFloat(total_ms.toFixed(4)),
  }),
);

File: benchmarks/pandas/bench_cum_ops.py
"""
Benchmark: Series.cumsum / cumprod / cummax / cummin / DataFrame.cumsum
Mirrors tsb bench_cum_ops.ts
"""
import json
import time
import pandas as pd

N = 100_000
WARMUP = 5
ITERS = 20

# Matching data
data = [(i % 100) + 1 for i in range(N)]
series = pd.Series(data, dtype=float)

col1 = [(i % 100) + 1 for i in range(N)]
col2 = [((i * 3) % 100) + 1 for i in range(N)]
df = pd.DataFrame({"a": col1, "b": col2}, dtype=float)

# Warm-up
for _ in range(WARMUP):
    series.cumsum()
    series.cummax()
    df.cumsum()

# Measured: cumsum
t0 = time.perf_counter()
for _ in range(ITERS):
    series.cumsum()
total_cumsum = (time.perf_counter() - t0) * 1000

# Measured: cumprod
t0 = time.perf_counter()
for _ in range(ITERS):
    series.cumprod()
total_cumprod = (time.perf_counter() - t0) * 1000

# Measured: cummax
t0 = time.perf_counter()
for _ in range(ITERS):
    series.cummax()
total_cummax = (time.perf_counter() - t0) * 1000

# Measured: cummin
t0 = time.perf_counter()
for _ in range(ITERS):
    series.cummin()
total_cummin = (time.perf_counter() - t0) * 1000

# Measured: DataFrame.cumsum
t0 = time.perf_counter()
for _ in range(ITERS):
    df.cumsum()
total_df = (time.perf_counter() - t0) * 1000

total_ms = total_cumsum + total_cumprod + total_cummax + total_cummin + total_df
mean_ms = total_ms / (ITERS * 5)

print(json.dumps({
    "function": "cum_ops",
    "mean_ms": round(mean_ms, 4),
    "iterations": ITERS * 5,
    "total_ms": round(total_ms, 4),
}))
